Store a copy of the input message if any stripping or truncation occurs

keszybz opened the pull request:
This allows the input stream to be recreated including any whitespace or parts that were after a zero byte and are not included in MESSAGE=. This replaces #9239, but is much smaller: no escaping is performed, only the _ORIGINAL_MESSAGE= field is added.

boucman commented:
should it be ORIGINAL_MESSAGE instead of _ORIGINAL_MESSAGE ?

it's clearly a user-provided information...

poettering commented on src/journal/journald-syslog.c:
```diff
@@ -304,44 +304,55 @@ void server_process_syslog_message(
         assert(s);
         assert(buf);
+        assert(buf_len > 0);
+        /* We know that buf is non-empty. It is also always NUL-terminated and can be used a string */
```
Hmm, these are weird calling conventions... passing a size *AND* requiring a NUL byte at the end is pretty uncommon. If we do this, then I think this deserves a bigger comment really, and it should clarify that the trailing NUL byte is there, but not counted in `buf_len`.

It might also be worth making these assumptions explicit, by adding `assert(buf[buf_len] == 0` or so.

Also, I think the name "buf_len" is a bit misleading now, given that the buffer is one byte larger than buf_len now. Maybe rename this to "raw" and "raw_len" or so?

keszybz replied:
I renamed buf_len to raw_len, and added a comment. This terminating byte allows strspn to be used, so I think it's worth keeping.

poettering commented on src/journal/journald-syslog.c:
```diff
@@ -304,44 +304,55 @@ void server_process_syslog_message(
                 forward_syslog_raw(s, priority, buf, buf_len, ucred, tv);
 
-        syslog_skip_date(&msg);
-        syslog_parse_identifier((const char**)&msg, &identifier, &pid);
```
moving this up is unnecessary, no?

keszybz replied:
Previous order was weird, but it had to be because the buffer was modified. Now it's not, so I just grouped all the related calls together.

poettering commented on src/journal/journald-syslog.c:
```diff
@@ -386,8 +391,18 @@ void server_process_syslog_message(
+        iovec[n++] = IOVEC_MAKE_STRING(message);
+
+        if (store_original) {
+                const size_t hlen = strlen("_ORIGINAL_MESSAGE=");
```
strlen() → STRLEN()

keszybz replied:
STRLEN is only useful to avoid VLAs, I don't think we need to use it otherwise.

poettering commented:
Hmm, it might be worth addressing #2398 at the same time, i.e. extend the condition under which ORIGINAL_MESSAGE is added to the case where the timestamp passed as part of the message doesn't match the kernel timestamp.

poettering commented:
> should it be ORIGINAL_MESSAGE instead of _ORIGINAL_MESSAGE ?
> 
> it's clearly a user-provided information...

That's true, the underscore should not be there. Also, I think this should contain SYSLOG or so in some form, matching SYSLOG_FACILITY=, SYSLOG_IDENTIFIER_= and SYSLOG_PID=. Maybe rename it SYSLOG_RAW= or so?

Also, this should be documented in systemd.journal-fields(7) with a brief mention why and when it is added, underlining that it is not always there.

keszybz commented:
```
Jun 20 09:24:21 autopkgtest systemd-udevd[1165]: Error resolving user 'usbmux': Connection timed out
Jun 20 09:24:46 autopkgtest systemd-udevd[1165]: Error resolving group 'render': Connection timed out
```
doesn't seem related, though it's suspicious that they all started failing now.

keszybz commented:
Updated. The new field is now called SYSLOG_RAW=, and is documented. I also added SYSLOG_TIMESTAMP= handle #2398. I first tried parsing the timestamp and including the SYSLOG_RAW= field if the parsed time is different than the (truncated) _SOURCE_REALTIME_TIMESTAMP= time, but this was ugly and complicated. In the end unconditionally including the timestamp in a separate field seems like a much nicer option.
